paasta_tools/autoscaling_lib.py

```diff
@@ -775,19 +778,18 @@ class ClusterAutoscalingError(Exception):
     pass
 
 
-def get_instance_id_from_ip(ip):
+def get_instance_id_from_ip(ip, instance_descriptions):
```

Comment:
>  Don't think this function should ever return more than one instance.

Sure, given the context, it makes sense that the length of the array returned is only 0  or 1. But,  if you do that, you're enforcing that opinion and the handling of the error on *everything* that calls the function. It might be in one place this only requires a ``log.warning``, but in another it requires it requires a ``ValueException`` or something else - you lose that flexibility if you put it in the function. 

>  It might return None if it can't find the IP you specify.

Nooo. This is just a wrapper around a ``filter``. Filter *always* just returns an empty list if none of the elements of the list passed to it 'match'. Please leave it that way.

If you also want to add a 'uniqueness' check, then do that separately and make it more generic (btw I totally see why you want this, just think the logic for it should be separate from the filtering). Maybe create a wrapper function if you think it'll be reused.

```
./paasta_tools/utils.py
def raise_unless_length(input_list, desired_length):
     if len(input_list) != desired_length:
         raise ValueError("Expected list of length %d, got %d" % (desired_length, len(input_list))
```

then just let this happen:

```
matching_instances = find_instance_for_ip(descriptions, ip)
raise_unless_length(matching_instances, 1)
```

I still think it's nasty because raise_unless_length is so side_effecty, but I guess it reduces the amount of code that's repeated in different places.